x-pack/plugin/apm-data: add dynamic setting for enabling template registry

axw opened the pull request:
Add `xpack.apm_data.registry.enabled` cluster setting, which can be used to disable the index template registry while restoring a snapshot.

Closes https://github.com/elastic/elasticsearch/issues/104385

Commit:
apm-data: add dynamic setting for registry

Add `xpack.apm_data.registry.enabled` cluster setting,
which can be used to disable the index template registry
while restoring a snapshot.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @axw, I've created a changelog YAML for you.

axw committed:
Update docs/changelog/104386.yaml

danielmitterdorfer reviewed (commented):
The code works fine; I left a few suggestions.

Commit:
Address review comments

- volatile instead of synchronized
- handle registry enablement setting in APMPlugin
- simpify createComponents and close:
  always create registry, never return it

axw commented on x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java:
```diff
@@ -76,7 +79,9 @@ public APMIndexTemplateRegistry(
 
-            enabled = XPackSettings.APM_DATA_ENABLED.get(nodeSettings);
+            setEnabled(APMPlugin.APM_DATA_REGISTRY_ENABLED.get(nodeSettings));
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(APMPlugin.APM_DATA_REGISTRY_ENABLED, this::setEnabled);
```
Good call, moved it there and introduced unit tests for APMPlugin to verify the setting behaviour.

axw commented on x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java:
```diff
@@ -33,13 +58,19 @@ public Collection<?> createComponents(PluginServices services) {
     @Override
     public void close() {
-        registry.get().close();
+        if (enabled) {
```
Done, thanks.

axw commented on x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java:
```diff
@@ -106,7 +116,7 @@ protected boolean requiresMasterNode() {
     @Override
     protected Map<String, ComponentTemplate> getComponentTemplateConfigs() {
-        if (enabled) {
+        if (isEnabled()) {
```
Done, thanks.

axw commented on x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java:
```diff
@@ -86,7 +91,12 @@ public int getVersion() {
     }
 
-    public boolean isEnabled() {
+    private synchronized void setEnabled(boolean enabled) {
```
Done, thanks.

danielmitterdorfer commented on x-pack/plugin/apm-data/src/test/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistryTests.java:
```diff
@@ -75,15 +77,23 @@ public class APMIndexTemplateRegistryTests extends ESTestCase {
+
         apmIndexTemplateRegistry = new APMIndexTemplateRegistry(
-            Settings.builder().put(APM_DATA_ENABLED.getKey(), true).build(),
+            // Settings.builder().put(APM_DATA_ENABLED.getKey(), true).put(APMPlugin.APM_DATA_REGISTRY_ENABLED.getKey(), true).build(),
```
Nit: Leftover?

danielmitterdorfer reviewed (commented):
The updated code looks fine to me. `APMIndexTemplateRegistryTests#testIngestPipelines()` is failing now though; supposedly the index template registry is disabled now but the test expects an enabled one.

Commit:
Fix tests; remove leftover

axw commented:
Thanks @danielmitterdorfer, apparently I forgot to run all the tests after making the latest change :man_facepalming:
Now explicitly enabling the registry in the `@Before`.

Commit:
Merge branch 'main' into apmdata-dynamic-registry

danielmitterdorfer commented on x-pack/plugin/apm-data/src/test/java/org/elasticsearch/xpack/apmdata/APMPluginTests.java:
```diff
@@ -0,0 +1,72 @@
+        threadPool.shutdownNow();
+    }
+
+    public void testRegistryEnabledSetting() throws Exception {
```
Just as a suggestion: If you want to be extra cautious you could also add a test to ensure the registry is disabled if the plugin is disabled.

danielmitterdorfer reviewed (approved):
LGTM. I left one optional suggestion.

Commit:
Test that disabling plugin disables registry

axw commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into apmdata-dynamic-registry

gmarouli commented on x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java:
```diff
@@ -10,36 +10,62 @@
-        return List.of(registryInstance);
+        if (enabled) {
+            APMIndexTemplateRegistry registryInstance = registry.get();
+            registryInstance.setEnabled(APM_DATA_REGISTRY_ENABLED.get(settings));
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(APM_DATA_REGISTRY_ENABLED, registryInstance::setEnabled);
```
We tend to encapsulate this in the code of the class using the `enabled` flag, in this case the `APMIndexTemplateRegistry`. We usually read the value of the setting and we set it in the constructor and in a method like `initialize()` we register the listener. This way it's transparent in the `APMIndexTemplateRegistry` how the enabled flag is changing.

axw replied:
@gmarouli is it worth me changing now? I can go back and update it.

The only way the `enabled` flag is updated is through the `setEnabled` method, and I'm not sure why `APMIndexTemplateRegistry` would care what caused that method to be called -- but if it's conventional and this difference will be confusing, I can go back and update it.

gmarouli replied:
It's more about encapsulation, the logic of how the flag for this class is changing is contained in the class itself. It's ok for now, if we come back to this we can address it.
